rpc/client.go

```diff
@@ -217,31 +219,35 @@ func (c *Client) Close() {
 		c.mu.Lock()
 		c.healthy = false
 		c.closed = true
-		c.mu.Unlock()
-		close(c.Closed)
 		if c.Client != nil {
 			c.Client.Close()
 		}
+		c.mu.Unlock()
+		close(c.Closed)
 	}
 	clientMu.Unlock()
 }
 
-// startHeartbeat sends periodic heartbeats to client. Closes the
+// runHeartbeat sends periodic heartbeats to client. Closes the
 // connection on error. Heartbeats are sent in an infinite loop until
 // an error is encountered.
-func (c *Client) startHeartbeat() {
+func (c *Client) runHeartbeat(stopper *util.Stopper) {
 	if log.V(2) {
 		log.Infof("client %s starting heartbeat", c.Addr())
 	}
-	// On heartbeat failure, remove this client from cache. A new
-	// client to this address will be created on the next call to
-	// NewClient().
+
 	for {
-		time.Sleep(heartbeatInterval)
-		if err := c.heartbeat(); err != nil {
-			log.Infof("client %s heartbeat failed: %v; recycling...", c.Addr(), err)
-			c.Close()
-			break
+		select {
+		case <-stopper.ShouldStop():
+			return
+		case <-time.After(heartbeatInterval):
+			// On heartbeat failure, remove this client from cache. A new
+			// client to this address will be created on the next call to
+			// NewClient().
```

Comment:
The comment is mildly confusing now that removal of the client from the cache happens because this method returns.